contracts/EventBasedPredictionMarket.sol

```diff
@@ -0,0 +1,333 @@
+     */
+    function settle(uint256 longTokensToRedeem, uint256 shortTokensToRedeem)
+        public
+        nonReentrant
+        returns (uint256 collateralReturned)
+    {
+        // Get the settlement price and store it. Also sets expiryPercentLong to inform settlement. Reverts if either:
+        // a) the price request has not resolved (either a normal expiration call or early expiration call) or b) If the
+        // the contract was attempted to be settled early but the price returned is the ignore oracle price.
+        // Note that we use the bool receivedSettlementPrice over checking for price != 0 as 0 is a valid price.
+        if (!receivedSettlementPrice) getExpirationPrice();
+
+        require(longToken.burnFrom(msg.sender, longTokensToRedeem));
+        require(shortToken.burnFrom(msg.sender, shortTokensToRedeem));
+
+        // expiryPercentLong is a number between 0 and 1e18. 0 means all collateral goes to short tokens and 1e18 means
+        // all collateral goes to the long token. Total collateral returned is the sum of payouts.
+        uint256 longCollateralRedeemed = FixedPoint
+            .Unsigned(longTokensToRedeem)
+            .mul(FixedPoint.Unsigned(collateralPerPair))
+            .mul(FixedPoint.Unsigned(expiryPercentLong))
+            .rawValue;
+        uint256 shortCollateralRedeemed = FixedPoint
+            .Unsigned(shortTokensToRedeem)
+            .mul(FixedPoint.Unsigned(collateralPerPair))
+            .mul(FixedPoint.fromUnscaledUint(1).sub(FixedPoint.Unsigned(expiryPercentLong)))
+            .rawValue;
+
+        collateralReturned = longCollateralRedeemed + shortCollateralRedeemed;
+        collateralToken.safeTransfer(msg.sender, collateralReturned);
+
+        emit PositionSettled(msg.sender, collateralReturned, longTokensToRedeem, shortTokensToRedeem);
+    }
+
+    /****************************************
+     *          INTERNAL FUNCTIONS          *
+     ****************************************/
+
+    // Return the oracle price for a given request timestamp and ancillary data combo.
+    function _getOraclePrice(uint256 requestTimestamp, bytes memory requestAncillaryData) internal returns (int256) {
+        return _getOptimisticOracle().settleAndGetPrice(priceIdentifier, requestTimestamp, requestAncillaryData);
+    }
+
+    // Fetch the optimistic oracle expiration price. If the oracle has the price for the provided expiration timestamp
+    // and customData combo then return this. If there is no price revert.
+    function getExpirationPrice() internal hasPrice {
+        expiryPrice = _getOraclePrice(expirationTimestamp, customAncillaryData);
+
+        // Finally, compute the value of expiryPercentLong based on the expiryPrice. Cap the return value at 1e18 as
+        // this should, by definition, between 0 and 1e18.
+        expiryPercentLong = Math.min(
+            financialProductLibrary.percentageLongCollateralAtExpiry(expiryPrice),
```

Comment:
so if you do what I recommended here: https://github.com/UMAprotocol/dev-quickstart/pull/1/files#r890408242 you can do this computation in-line